src/slang-utils/create-parser.ts

```diff
@@ -0,0 +1,90 @@
+import { NonterminalKind, Query } from '@nomicfoundation/slang/cst';
+import { Parser } from '@nomicfoundation/slang/parser';
+import strip from 'strip-comments';
+import {
+  maxSatisfying,
+  minSatisfying,
+  minor,
+  major,
+  satisfies,
+  validRange
+} from 'semver';
+
+const supportedVersions = Parser.supportedVersions();
+
+const milestoneVersions = Array.from(
+  supportedVersions.reduce((minorRanges, version) => {
+    minorRanges.add(`^${major(version)}.${minor(version)}.0`);
+    return minorRanges;
+  }, new Set<string>())
+)
+  .reverse()
+  .reduce((versions: string[], range) => {
+    versions.push(maxSatisfying(supportedVersions, range)!);
+    versions.push(minSatisfying(supportedVersions, range)!);
+    return versions;
+  }, []);
+
+// TODO if we ended up selecting the same version that the pragmas were parsed with,
+// should we be able to reuse/just return the already parsed CST, instead of
+// returning a Parser and forcing user to parse it again?
+export function createParser(text: string): Parser {
+  let inferredRanges: string[] = [];
+
+  for (const version of milestoneVersions) {
+    try {
+      inferredRanges = tryToCollectPragmas(text, version);
+      break;
+    } catch {}
+  }
+
+  const satisfyingVersions = inferredRanges.reduce(
+    (versions, inferredRange) => {
+      if (!validRange(inferredRange)) {
+        throw new Error(
+          "Couldn't infer any version from the ranges in the pragmas."
+        );
+      }
+      return versions.filter((supportedVersion) =>
+        satisfies(supportedVersion, inferredRange)
+      );
+    },
+    supportedVersions
+  );
+
+  return satisfyingVersions.length > 0
+    ? Parser.create(satisfyingVersions[satisfyingVersions.length - 1])
+    : Parser.create(supportedVersions[supportedVersions.length - 1]);
+}
+
+function tryToCollectPragmas(text: string, version: string): string[] {
+  const language = Parser.create(version);
```

Comment:
suggestion to remove the need for the extra dependency, and to be able to handle things like newlines/other corner cases:

```ts
const fileCursor = parser.parse(NonterminalKind.SourceUnit, source).createTreeCursor();
const values = [];

while (fileCursor.goToNextNonterminalWithKind(NonterminalKind.VersionExpressionSets)) {
  const versionCursor = fileCursor.spawn();
  let value = "";

  while (versionCursor.goToNextTerminal()) {
    const terminal = versionCursor.node.asTerminalNode()!;
    if (TerminalKindExtensions.isTrivia(terminal.kind)) {
      value += " "; // to break up any possible consecutive identifiers
    } else {
      value += terminal.unparse();
    }
  }

  values.push(value);
}
```

Reply:
I tried using this and tests aren't passing. I suspect is because we might be moving through an AST with errors when the syntax doesn't match the parser's version.

Reply:
That syntax should be consistent across all versions. I wonder if you have a commit # I can try out?
Happy to help if I can debug the failure..